packages/@aws-cdk/aws-ec2/lib/security-group.ts

```diff
@@ -85,6 +127,75 @@ export abstract class SecurityGroupBase extends Construct implements ISecurityGr
   public abstract export(): SecurityGroupImportProps;
 }
 
+/**
+ * Determine where to parent a new ingress/egress rule
+ *
+ * A SecurityGroup rule is parented under the group it's related to, UNLESS
+ * we're in a cross-stack scenario with another Security Group. In that case,
+ * we respect the 'remoteRule' flag and will parent under the other security
+ * group.
+ *
+ * This is necessary to avoid cyclic dependencies between stacks, since both
+ * ingress and egress rules will reference both security groups, and a naive
+ * parenting will lead to the following situation:
+ *
+ *   ╔════════════════════╗         ╔════════════════════╗
+ *   ║  ┌───────────┐     ║         ║    ┌───────────┐   ║
+ *   ║  │  GroupA   │◀────╬─┐   ┌───╬───▶│  GroupB   │   ║
+ *   ║  └───────────┘     ║ │   │   ║    └───────────┘   ║
+ *   ║        ▲           ║ │   │   ║          ▲         ║
+ *   ║        │           ║ │   │   ║          │         ║
+ *   ║        │           ║ │   │   ║          │         ║
+ *   ║  ┌───────────┐     ║ └───┼───╬────┌───────────┐   ║
+ *   ║  │  EgressA  │─────╬─────┘   ║    │ IngressB  │   ║
+ *   ║  └───────────┘     ║         ║    └───────────┘   ║
+ *   ║                    ║         ║                    ║
+ *   ╚════════════════════╝         ╚════════════════════╝
+ *
+ * By having the ability to switch the parent, we avoid the cyclic reference by
+ * keeping all rules in a single stack.
+ *
+ * If this happens, we also have to change the construct ID, because
+ * otherwise we might have two objects with the same ID if we have
+ * multiple reversed security group relationships.
+ *
+ *   ╔═══════════════════════════════════╗
+ *   ║┌───────────┐                      ║
+ *   ║│  GroupB   │                      ║
+ *   ║└───────────┘                      ║
+ *   ║      ▲                            ║
+ *   ║      │              ┌───────────┐ ║
+ *   ║      ├────"from A"──│ IngressB  │ ║
+ *   ║      │              └───────────┘ ║
+ *   ║      │              ┌───────────┐ ║
+ *   ║      ├─────"to B"───│  EgressA  │ ║
+ *   ║      │              └───────────┘ ║
+ *   ║      │              ┌───────────┐ ║
+ *   ║      └─────"to B"───│  EgressC  │ ║  <-- oops
+ *   ║                     └───────────┘ ║
+ *   ╚═══════════════════════════════════╝
+ */
+function determineRuleParent(
+      group: SecurityGroupBase,
+      peer: ISecurityGroupRule,
+      connection: IPortRange,
+      fromTo: 'from' | 'to',
+      remoteRule?: boolean): [SecurityGroupBase, string] {
+
+  if (remoteRule && SecurityGroupBase.isSecurityGroup(peer) && differentStacks(group, peer)) {
+    // Reversed
+    const reversedFromTo = fromTo === 'from' ? 'to' : 'from';
+    return [peer, `${group.uniqueId}:${connection} ${reversedFromTo}`.replace('/', '_')];
```

Comment:
BTW, you don't need to replace "/"s, they will automatically be replaced when used as IDs.

Reply:
This didn't use to be true, but indeed :)

Reply:
I know, it's a shiny new thing